zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+    if (files == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get file info.");
+    }
+
+    // Read the file info and add to the list. If an exception is thrown, the entire operation will fail
+    List<BackupFileInfo> backupFileInfos = new ArrayList<>();
+    for (File file : files) {
+      BasicFileAttributes fileAttributes =
+          Files.readAttributes(Paths.get(file.getPath()), BasicFileAttributes.class);
+      backupFileInfos.add(new BackupFileInfo(file, fileAttributes.lastModifiedTime().toMillis(),
+          fileAttributes.size()));
+    }
+    return backupFileInfos;
+  }
+
+  @Override
+  public List<File> getDirectories(File path) throws IOException {
+    String backupDirPath = constructBackupFilePath(path.getPath());
+    File backupDir = new File(backupDirPath);
+
+    if (!backupDir.exists()) {
+      throw new BackupException(
+          "Backup directory " + path.getPath() + " does not exist, could not get directory list.");
+    }
+
+    // Filter out all the files which are directories
+    FilenameFilter fileFilter = (dir, name) -> new File(dir, name).isDirectory();
+    File[] dirs = backupDir.listFiles(fileFilter);
+
+    if (dirs == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get directory list.");
+    }
+    return Arrays.asList(dirs);
+  }
+
+  @Override
+  public void copyToBackupStorage(File srcFile, File destName) throws IOException {
+    InputStream inputStream;
+    OutputStream outputStream = null;
+
+    // Create input stream from the source file
+    try {
+      inputStream = new FileInputStream(srcFile);
+    } catch (FileNotFoundException e) {
+      if (!srcFile.exists() || srcFile.isDirectory()) {
+        throw new BackupException("File " + srcFile.getPath()
+            + " does not exist or is a directory, cannot be copied to backup storage.");
+      } else {
+        throw new IOException("Cannot open source file " + srcFile.getPath()
+            + " , cannot be copied to backup storage.");
+      }
+    }
+
+    // Prepare in the backup storage: create parent dirs if not exist, clean up temporary files under the parent dir
+    String backupTempFilePath = constructBackupFilePath(constructTempFileName(destName.getName()));
+    String backupFilePath = constructBackupFilePath(destName.getName());
+    File backupTempFile = new File(backupTempFilePath);
+    File backupFile = new File(backupFilePath);
+    File parentDir = backupTempFile.getParentFile();
+    parentDir.mkdirs();
+    cleanupTempFiles(parentDir.getPath());
+
+    // Create destination temporary file
+    if (!backupTempFile.createNewFile()) {
+      if (backupFile.exists()) {
+        throw new BackupException("A backed up file with the file path " + backupFilePath
+            + " already exists, could not create a new one.");
```

Comment:
Could you explain why we're checking for the existence of `backupFile`, not `backupTempFile` here?

Reply:
+1. What if it is leftover work from last start up? I remember it should be the file will be override, right? We should allow even the backup file does exits

Reply:
Has this been resolved?